integration-tests/smoke/ccip/ccip_reorg_test.go

```diff
@@ -0,0 +1,556 @@
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/onsi/gomega"
+	chainsel "github.com/smartcontractkit/chain-selectors"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/zapcore"
+
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/utils/tests"
+
+	"github.com/smartcontractkit/chainlink-ccip/pkg/types/ccipocr3"
+
+	"github.com/smartcontractkit/chainlink-evm/gethwrappers/ccip/generated/v1_2_0/router"
+	"github.com/smartcontractkit/chainlink-evm/gethwrappers/ccip/generated/v1_6_0/onramp"
+	ctf_client "github.com/smartcontractkit/chainlink-testing-framework/lib/client"
+	"github.com/smartcontractkit/chainlink-testing-framework/lib/logging"
+	"github.com/smartcontractkit/chainlink/deployment"
+	ccipcs "github.com/smartcontractkit/chainlink/deployment/ccip/changeset"
+	"github.com/smartcontractkit/chainlink/deployment/ccip/changeset/testhelpers"
+	"github.com/smartcontractkit/chainlink/deployment/environment/nodeclient"
+	testsetups "github.com/smartcontractkit/chainlink/integration-tests/testsetups/ccip"
+)
+
+var (
+	logsToIgnoreOpt = testhelpers.WithLogMessagesToIgnore([]testhelpers.LogMessageToIgnore{
+		{
+			Msg:    "Got very old block.",
+			Reason: "We are expecting a re-org beyond finality",
+			Level:  zapcore.DPanicLevel,
+		},
+		{
+			Msg:    "Reorg greater than finality depth detected",
+			Reason: "We are expecting a re-org beyond finality",
+			Level:  zapcore.DPanicLevel,
+		},
+		{
+			Msg:    "Failed to poll and save logs due to finality violation, retrying later",
+			Reason: "We are expecting a re-org beyond finality",
+			Level:  zapcore.DPanicLevel,
+		},
+	})
+)
+
+const (
+	greaterThanFinalityReorgDepth = 60
+	lessThanFinalityReorgDepth    = 7
+)
+
+func Test_CCIPReorg_BelowFinality_OnSource(t *testing.T) {
+	e, l, dockerEnv, nonBootstrapP2PIDs, state := setupReorgTest(t,
+		testhelpers.WithExtraConfigTomls([]string{t.Name() + ".toml"}),
+	)
+
+	// Chain setup
+	allChains := e.Env.AllChainSelectors()
+	require.GreaterOrEqual(t, len(allChains), 2)
+	sourceSelector := allChains[0]
+	destSelector := allChains[1]
+
+	// Build RPC map and get clients
+	chainSelToRPCURL := buildChainSelectorToRPCURLMap(t, dockerEnv)
+	sourceClient := ctf_client.NewRPCClient(chainSelToRPCURL[sourceSelector], nil)
+
+	// Setup CCIP lane
+	testhelpers.AddLaneWithDefaultPricesAndFeeQuoterConfig(t, &e, state, sourceSelector, destSelector, false)
+	waitForLogPollerFilters(l)
+
+	// Send initial message
+	msgBeforeReorg := sendCCIPMessage(t, e.Env, state, sourceSelector, destSelector, l)
+
+	// Wait and perform reorg
+	minBlock := msgBeforeReorg.Raw.BlockNumber + lessThanFinalityReorgDepth - 1
+	waitForBlockNumber(t, sourceClient, minBlock, 1*time.Minute, 500*time.Millisecond, l)
+	performReorg(t, sourceClient, lessThanFinalityReorgDepth, l)
```

Comment:
nit, but how about performing multiple reorgs instead of only one? I'm thinking about something like this

```go
finalityDepth = 50

for i := 0; i < reorgAttempts; i++ {
    performReorg(10 blocks)
    emit couple of logs
}```

Reply:
Performing multiple reorgs on different chains or on the same chain?